### DISCUSSION TITLE
google/gsuite: missing refresh token errors if auth request params set

### DISCUSSION THREAD

**Discussion by whs** [pomerium/pomerium#1464] [General] [answered]
## What happened?
After running changes in #1448 , users get logged out every 30 min.

I believe it is due to this error

```json
{"level":"error","service":"identity_manager","error":"identity/oidc: missing refresh token","user_id":"google/xxxxx","session_id":"62efd788-f4c4-48cd-9d6e-14f01a77c2b9","time":"2020-09-25T07:21:30Z","message":"failed to refresh oauth2 token, deleting session"}
{"level":"error","service":"identity_manager","error":"identity/oidc: user info endpoint: 401 Unauthorized: {\"error\":\"invalid_request\",\"error_description\":\"Invalid Credentials\"}","user_id":"google/xxxxx","session_id":"62efd788-f4c4-48cd-9d6e-14f01a77c2b9","time":"2020-09-25T07:21:30Z","message":"failed to update user info, deleting session"}
```

Pomerium-redirected Google login URL: https://accounts.google.com/o/oauth2/v2/auth/identifier?access_type=offline&client_id=xxxxx&hd=lmwn.com&redirect_uri=https%3A%2F%2Fxxxxx%2Foauth2%2Fcallback&response_type=code&scope=openid%20profile%20email&state=xxxx&flowName=GeneralOAuthFlow

I added hd in configuration, but I believe all the necessary parameters are still present? (access_type=offline, scope is still default value)

## What's your environment like?

- Pomerium version (retrieve with `pomerium --version`): https://github.com/wongnai/pomerium master (which is pomerium/master f4c61a0cdc9456574cc3f4b44105bfd08b2b9703 plus my two PRs)
- Server Operating System/Architecture/Cloud: Debian buster from Docker Hub

## What's your config.yaml?
Please see #1448

**Comment by desimone** [comment, marked as answer]
Hey @whs ,

I believe this is due to a (not well documented) quirk with Google's OpenID Connect implementation. When you are setting the hd value, it is replacing the default auth request params which if unset do not allow offline access type to be set, even if specified. 

https://github.com/pomerium/pomerium/blob/1e7723f0d29c8d9b72742a0ebe0cca70bd1ae04c/internal/identity/oidc/google/google.go#L27

Google only provide refresh_token on the first authorization from the user. If user clears  cookies, re-authorization will not bring back refresh_token. A work around to this is to add  prompt=consent to the OAuth redirect URL and will always return a refresh_token. 

I've definitely been tripped up by this before. 


Related:
- https://stackoverflow.com/questions/43567735/how-to-get-offline-access-in-google-api-without-prompting-the-user
- https://openid.net/specs/openid-connect-core-1_0.html#OfflineAccess 
- https://github.com/pomerium/pomerium/blob/712d02c3ccb74a46025ea1f505d2401c2d0ce8e6/internal/identity/google.go#L113-L124

**Comment by whs** [comment]
That definitely helped. Thanks!

**Comment by desimone** [comment]
(converted to discussion)
